Make ccall return Promises with ASYNCIFY WASM Backend (Fixes #9029)

AjayP13 opened the pull request:
Addresses Issue #9020.

It mostly works right now, probably need a little bit help to get it complete.

Working:

- Removes warning assertion (#9124)
- Adds warning for async execution in ccall without `opts.async`.
- Adds warning if attempting multiple in-flight async executions.
- Adds `asyncFinalizers` to `Asyncify` that are called when all asynchronous execution activity is completed.

Not working:
- Passes `restartFunc` to `Asyncify` which re-calls the function before calling `asyncFinalizers` with the return value from the `restartFunc`. This actually seems to work sometimes to get the correct return value to the Promise `resolve()`, but it is not consistently working. I mostly ported this over from the Emterpreter Asyncify implementation, so this part might be wrong. If so, what is the appropriate way to get the final return value after all asynchronous execution has completed?

AjayP13 commented on src/library_async.js:
```diff
@@ -671,7 +672,14 @@ mergeInto(LibraryManager.library, {
 #endif
-          Module['asm'][start]();
+          var asyncReturnValue = Module['asm'][start]();
+          if (Asyncify.currData === null) {
```
`(Asyncify.currData === null)` means that no asynchronous execution is running.

kripken replied:
In other places we just do `!Asyncify.currData` which is shorter.

AjayP13 replied:
Sounds good to me, I'll update that. I was worried about whether `currData` could be 0 since it is an integer type, but I suppose it is probably never 0 as it is a stack pointer.

AjayP13 commented on src/preamble.js:
```diff
@@ -130,14 +130,23 @@ function ccall(ident, returnType, argTypes, args, opts) {
     });
   }
 #else // EMTERPRETIFY_ASYNC
+  if (typeof Asyncify === 'object' && Asyncify.currData !== null) {
```
`Asyncify.currData !== null` means that asynchronous execution is currently happening.

AjayP13 commented on src/preamble.js:
```diff
@@ -130,14 +130,23 @@ function ccall(ident, returnType, argTypes, args, opts) {
 #if ASSERTIONS
-  assert(!(opts && opts.async), 'async call is only supported with Emterpretify for now, see #9029');
+    assert(opts && opts.async, 'The call to ' + ident + ' is running asynchronously. If this was intended, add the async option to the ccall/cwrap call.');
+    assert(Asyncify.asyncFinalizers.length === 0, 'Cannot have multiple async ccalls in flight at once');
```
If the assertion for `Asyncify.asyncFinalizers.length === 0` fails, that means that another `async ccall` is currently in-flight and waiting to have it's finalizers called, since once the finalizers are called, `asyncFinalizers` gets set back to `[]`.

kripken replied:
I think this an other comments in the PR could be comments in the source. Then they won't be lost later.

AjayP13 commented on src/preamble.js:
```diff
@@ -130,14 +130,23 @@ function ccall(ident, returnType, argTypes, args, opts) {
   ret = convertReturnValue(ret);
   if (stack !== 0) stackRestore(stack);
-#if EMTERPRETIFY_ASYNC
+#if EMTERPRETIFY_ASYNC || ASYNCIFY
```
We can directly return a resolved Promise if run in `async` mode, but no `Asyncify.handleSleep` ever occurred.

AjayP13 commented on src/library_async.js:
```diff
@@ -671,7 +672,14 @@ mergeInto(LibraryManager.library, {
           err('ASYNCIFY: start: ' + start);
 #endif
-          Module['asm'][start]();
+          var asyncReturnValue = Module['asm'][start]();
```
`Module['asm'][start]()` returns the return value on it's final call once, all asynchronous execution has completed.

kripken replied:
It's unclear to me how asyncReturnValue differs from returnValue from earlier above, which is returned from handleSleep?

AjayP13 replied:
I believe it is a different return value. I tried returning that one and it always seemed to hold the value `0` (its default value). It seems to be for this return value: https://emscripten.org/docs/porting/asyncify.html#returning-values

i.e. `Asyncify.returnValue` holds the optional return value you return from `asynchronous JS ----> C` using `wakeUp(<contents of returnValue>)`. Then, `Asyncify.returnValue` becomes the return value of the C function in [`ASYNCIFY_IMPORTS`](https://emscripten.org/docs/porting/asyncify.html#more-on-asyncify-imports).  For example on [this example](https://emscripten.org/docs/porting/asyncify.html#returning-values), it becomes the return value of the C function `get_digest_size()`. Eventually, `Asyncify.returnValue` becomes the return value of the **last** C function in [`ASYNCIFY_IMPORTS`](https://emscripten.org/docs/porting/asyncify.html#more-on-asyncify-imports) to be executed. 

However, this is **not** the same as the return value of the function being `ccall`-ed. The trivial example is the following:

```c
int foo() {
     printf("Before Sleep\n");
     emscripten_sleep(1000);
     printf("After Sleep\n");
     return 1337;
}
```

If `foo` is being `ccall`-ed, we should return `1337`. However, if you return `Asyncify.returnValue` it will be the return value of `emscripten_sleep`, or `0` since it is `void`. This is because `emscripten_sleep` is the function in `ASYNCIFY_IMPORTS` and it is defined like so:
```js
  emscripten_sleep: function(ms) {
    Asyncify.handleSleep(function(wakeUp) {
      Browser.safeSetTimeout(wakeUp, ms);
    });
  },
```
When `setTimeout` calls `wake` with no arguments (`wakeUp(<VOID NO RETURN VALUE>)`), so `Asyncify.returnValue` remains it's old, default value of 0.


On the other hand, `asyncReturnValue` will contain the return value of `foo`. When a WASM function is asyncified, it is called `N` times by `Module['asm'][start]()` to restart the function execution, if there are `N` unwindings of the stack during the C execution. On the final call of `Module['asm'][start]()`, it will actually have executed it's `return` executed and will return its return value from the WASM function the JavaScript code which we capture into `asyncReturnValue` and return to the `asyncFinalizers`.

AjayP13 replied:
I agree this is confusing though due to this comment:

```js
    // The return value from a synchronous call is stored here,
    // so we can return it later after rewinding finishes.
    returnValue: 0,
```

It is the return value of the synchronous call of the function in `ASYNC_IMPORTS` (`emscripten_sleep`), not the return value of a function that has called `emscripten_sleep` sleep like `foo`.

AjayP13 replied:
Added comments explaining the differences to the source file, but perhaps could even benefit from renaming of the variables. I'm thinking:

`Asyncify.returnValue` ---> `handleSleepReturnValue`
`asyncReturnValue` ---> `asyncWasmReturnValue`

AjayP13 commented on tests/test_core.py:
```diff
@@ -7299,7 +7299,7 @@ def test_async(self, emterpretify=False):
     self.do_run(src, 'HelloWorld!99')
 
-    if emterpretify:
+    if self.is_wasm_backend() or emterpretify:
```
We can now run these tests on the WASM backend to ensure `ccall` use works correctly.

AjayP13 commented:
@kripken I added some comments around the changes so you understand what's changed and enabled the async `ccall` test cases to run for the LLVM WASM backend Asyncify code path (see `test-upstream-wasm0`). 

Should be ready to merge once you are able to take a look at this. :)

AjayP13 commented:
Added comments to the JS files as well as comments in `library_async.js` about the difference between `Asyncify.returnValue` and `asyncReturnValue`.

AjayP13 commented:
Added the proper #ifdefs and renamed the return value variables to make it
clearer.

On Mon, Aug 26, 2019, 3:36 PM Alon Zakai <notifications@github.com> wrote:

> *@kripken* commented on this pull request.
>
> Great! Looks good, just a few final comments.
> ------------------------------
>
> In src/library_async.js
> <https://github.com/emscripten-core/emscripten/pull/9280#discussion_r317824349>
> :
>
> > @@ -671,7 +672,14 @@ mergeInto(LibraryManager.library, {
>  #if ASYNCIFY_DEBUG
>            err('ASYNCIFY: start: ' + start);
>  #endif
> -          Module['asm'][start]();
> +          var asyncReturnValue = Module['asm'][start]();
>
> I see, thanks @AjayP13 <https://github.com/AjayP13>! (Sorry for the
> delay, I was on vacation...)
>
> I agree with the renaming, let's change it to that. Indeed once is the
> return value for handleSleep, and the other for the the entire wasm.
> ------------------------------
>
> In src/preamble.js
> <https://github.com/emscripten-core/emscripten/pull/9280#discussion_r317825392>
> :
>
> >  #if ASSERTIONS
> -  assert(!(opts && opts.async), 'async call is only supported with Emterpretify for now, see #9029');
> +    assert(opts && opts.async, 'The call to ' + ident + ' is running asynchronously. If this was intended, add the async option to the ccall/cwrap call.');
> +    // Once the asyncFinalizers are called, asyncFinalizers gets reset to [].
> +    // If they are not empty, then another async ccall is in-flight and not finished.
> +    assert(Asyncify.asyncFinalizers.length === 0, 'Cannot have multiple async ccalls in flight at once');
>
> please put the asyncify code behind an ifdef for #if ASYNCIFY &&
> WASM_BACKEND, so that it does not appear when not using the feature
> ------------------------------
>
> In src/preamble.js
> <https://github.com/emscripten-core/emscripten/pull/9280#discussion_r317826492>
> :
>
> >  #endif // EMTERPRETIFY_ASYNC
>
>    ret = convertReturnValue(ret);
>    if (stack !== 0) stackRestore(stack);
> -#if EMTERPRETIFY_ASYNC
> +#if EMTERPRETIFY_ASYNC || ASYNCIFY
>
> this should be ASYNCIFY && WASM_BACKEND (as the old asyncify with
> fastcomp does not support this)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/emscripten-core/emscripten/pull/9280?email_source=notifications&email_token=ABJHMENI6THJUXE5WVPGZ7DQGRLITA5CNFSM4IOCI4D2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCCXHYLA#pullrequestreview-279870508>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABJHMEKG3JQDQCFBYNTJSG3QGRLITANCNFSM4IOCI4DQ>
> .
>
